File: src/pratice/binarysearch/공유기설치2110.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 공유기설치2110 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, C;
    public static int[] distances;
    public static void main(String[] args) throws IOException {
        input();
        Arrays.sort(distances);

        int L = 1;
        int R = distances[N-1] - distances[0]; // 최대 거리
        int res = L;
        while(L <= R){
            int mid = (L + R) / 2;
            if(canSetUp(mid)){
                res = mid;
                L = mid + 1;
            }else{
                R = mid - 1;
            }
        }
        System.out.println(res);
    }

    private static boolean canSetUp(int mid) {
        int cnt = 1;
        int last = distances[0];
        for (int i = 1; i < N; i++) {
           if(distances[i] - last >= mid){
               cnt++;
               last = distances[i];
           }
            if(cnt >= C){
                return true;
            }
        }
        return false;
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        C = Integer.parseInt(split[1]);
        distances = new int[N];
        for (int i = 0; i < N; i++) {
            distances[i] = Integer.parseInt(br.readLine());
        }
    }
}

File: src/pratice/binarysearch/공유기설치2110.kt
package pratice.binarysearch

fun main() {
    val (N, M) = readln().split(" ").map { it.toInt() }
    val arr = mutableListOf<Int>()
    repeat(N) {
        arr.add(readln().toInt())
    }
    arr.sort()
    val res = lowerBound(arr, M)
    println(res)
}
private fun lowerBound(arr: MutableList<Int>, M: Int): Int {
    var L = 1
    var R = arr[arr.size - 1] - arr[0]
    var res = L

    while (L <= R) {
        val mid = (L + R) / 2
        if (useAble(arr, mid, M)) {
            res = mid
            L = mid + 1
        } else {
            R = mid - 1
        }
    }
    return res
}

fun useAble(arr: MutableList<Int>, x: Int, M: Int): Boolean {
    var cnt = 1
    var last = arr[0]
    for (i in 1 until arr.size) {
        if (arr[i] - last >= x) {
            cnt++
            last = arr[i]
        }
        if (cnt >= M) {
            return true
        }
    }
    return false
}
